### PR TITLE
Replace delayed/pause admins with Guardian contract

### PR DESCRIPTION
**Pull request by AStox**
This PR replaced the delayedAdmin and pauseAdmin with a new Guardian contract, which handles the responsibilities of both delayedAdmin and PauseAdmin, but uses a Gnosis Safe to handle wards. 

New Auth flow:
The Gnosis safe, `AdminSafe`, can schedule/cancel relys, and unpause
The Gnosis safe owners, can pause.

Adding/removing owners is handled via Gnosis Safe's UI.

### CODE REVIEW

File: script/Deployer.sol [hunk 1]
```diff
@@ -80,14 +75,15 @@ contract Deployer is Script {
...
         gateway.rely(address(aggregator));
 
-        // Wire admins
-        pauseAdmin.rely(address(delayedAdmin));
-        root.rely(address(pauseAdmin));
-        root.rely(address(delayedAdmin));
-        root.rely(address(gateway));
-        aggregator.rely(address(delayedAdmin));
+        // Deploy gateway and guardian
+        guardian = new Guardian(address(root), adminSafe, address(aggregator));
```
Comment (hieronx) [inline script/Deployer.sol:79]: Why isn't the deployment happening still in the `deploy` method?

File: test/mocks/MockSafe.sol [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.21;
+
+import "./Mock.sol";
+
+contract MockSafe is Mock {
```
Comment (hieronx) [inline test/mocks/MockSafe.sol:6]: Nice mock!

File: test/integration/Deploy.t.sol [hunk 3]
```diff
@@ -62,25 +66,19 @@ contract DeployTest is Test, Deployer {
...
 
-        assertEq(delayedAdmin.wards(admin), 1);
-        assertEq(delayedAdmin.wards(nonAdmin), 0);
-
-        assertEq(pauseAdmin.pausers(pausers[0]), 1);
-        assertEq(pauseAdmin.pausers(pausers[1]), 1);
-        assertEq(pauseAdmin.pausers(pausers[2]), 1);
-        assertEq(pauseAdmin.pausers(nonPauser), 0);
+        assertEq(address(guardian.safe()), adminSafe);
+        assertEq(MockSafe(adminSafe).isOwner(pausers[0]), true);
```
Comment (hieronx) [inline test/integration/Deploy.t.sol:80]: Shouldn't this loop over `pausers`?

File: src/admin/Guardian.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+/// @dev    This contract allows a Gnosis Safe to schedule and cancel new relys,
+///         and unpause the protocol through the timelock of Root. Additionally,
+///         it allows any owners of the safe to instantly pause the protocol.
+
+contract Guardian {
+    Root public immutable root;
+    SafeLike public immutable safe;
+    RouterAggregatorLike public immutable aggregator;
+
+    constructor(address root_, address safe_, address aggregator_) {
```
Comment (hieronx) [inline src/admin/Guardian.sol:26]: I'd swap this, as root and aggregator are the external dependencies so should fit together
```suggestion
    constructor(address safe_, address root_, address aggregator_) {
```

File: src/admin/Guardian.sol [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+interface RouterAggregatorLike {
+    function disputeMessageRecovery(bytes32 messageHash) external;
+}
+
+/// @title  Guardian
+/// @dev    This contract allows a Gnosis Safe to schedule and cancel new relys,
+///         and unpause the protocol through the timelock of Root. Additionally,
+///         it allows any owners of the safe to instantly pause the protocol.
+
+contract Guardian {
```
Comment (hieronx) [inline src/admin/Guardian.sol:21]: Should inherit from an `IGuardian` interface now

**Review by hieronx — COMMENTED** [review]
Nice work 🔥

File: src/admin/Guardian.sol [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        aggregator = RouterAggregatorLike(aggregator_);
+    }
+
+    modifier onlySafe() {
+        require(msg.sender == address(safe), "Guardian/not-an-authorized-safe");
+        _;
+    }
+
+    modifier onlyOwner() {
+        require(_isSafeOwner(safe, msg.sender), "Guardian/not-an-owner-of-the-authorized-safe");
```
Comment (hieronx) [inline src/admin/Guardian.sol:38]: I'd change this to:
```suggestion
    modifier onlySafeOrOwner() {
        require(msg.sender == address(safe) || _isSafeOwner(safe, msg.sender), "Guardian/not-an-owner-of-the-authorized-safe");
```

Why: this way, we can also use the entire guardian functionalities without an actual Gnosis Safe deployment (where `safe` is maybe another multisig, or just an EOA). This is especially helpful on testnet where there might not always be Gnosis Safe support

File: src/admin/Guardian.sol [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    modifier onlySafe() {
+        require(msg.sender == address(safe), "Guardian/not-an-authorized-safe");
+        _;
+    }
+
+    modifier onlySafeOrOwner() {
+        require(
+            msg.sender == address(safe) || _isSafeOwner(safe, msg.sender),
```
Comment (hieronx) [inline src/admin/Guardian.sol:40]: No need to pass the `safe` here, `_isSafeOwner` can just query the variable
```suggestion
            msg.sender == address(safe) || _isSafeOwner(msg.sender),
```

File: src/admin/Guardian.sol [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+interface RouterAggregatorLike {
+    function disputeMessageRecovery(bytes32 messageHash) external;
+}
+
+/// @title  Guardian
+/// @dev    This contract allows a Gnosis Safe to schedule and cancel new relys,
+///         and unpause the protocol through the timelock of Root. Additionally,
+///         it allows any owners of the safe to instantly pause the protocol.
+
```
Comment (hieronx) [inline src/admin/Guardian.sol:21]:
```suggestion
```

File: src/admin/Guardian.sol [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import {Root} from "src/Root.sol";
+import {IGuardian} from "src/interfaces/IGuardian.sol";
+
+interface SafeLike {
+    function getOwners() external view returns (address[] memory);
+    function isOwner(address signer) external view returns (bool);
+    function getThreshold() external view returns (uint256);
+}
+
+interface RouterAggregatorLike {
```
Comment (hieronx) [inline src/admin/Guardian.sol:13]:
```suggestion
interface AggregatorLike {
```

File: src/admin/Guardian.sol [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+/// @title  Guardian
+/// @dev    This contract allows a Gnosis Safe to schedule and cancel new relys,
+///         and unpause the protocol through the timelock of Root. Additionally,
+///         it allows any owners of the safe to instantly pause the protocol.
+
+contract Guardian is IGuardian {
+    Root public immutable root;
+    SafeLike public immutable safe;
+    RouterAggregatorLike public immutable aggregator;
```
Comment (hieronx) [inline src/admin/Guardian.sol:25]:
```suggestion
    AggregatorLike public immutable aggregator;
```

File: src/admin/Guardian.sol [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+    constructor(address safe_, address root_, address aggregator_) {
+        root = Root(root_);
+        safe = SafeLike(safe_);
+        aggregator = RouterAggregatorLike(aggregator_);
```
Comment (hieronx) [inline src/admin/Guardian.sol:30]:
```suggestion
        aggregator = AggregatorLike(aggregator_);
```

**Review by hieronx — COMMENTED** [review]
A few nitpicks, but besides that it's looking good!

**Review by hieronx — APPROVED** [review]

File: script/Deployer.sol [hunk 12]
```diff
@@ -123,7 +109,5 @@ contract Deployer is Script {
         escrow.deny(deployer);
         gateway.deny(deployer);
         aggregator.deny(deployer);
-        pauseAdmin.deny(deployer);
-        delayedAdmin.deny(deployer);
```
Comment (ilinzweilin) [inline script/Deployer.sol:127]: guardian has to deny the deployer as well

**Review by ilinzweilin — CHANGES_REQUESTED** [review]

File: script/Deployer.sol (hunk 12 above)
Comment (AStox) [reply script/Deployer.sol:127]: Guardian.sol doesn't inherit auth, so no need to deny here

**Review by ilinzweilin — APPROVED** [review]

**Pull request merged**